Added show_message_box

Cobrand opened the pull request:
- Added show_message_box in the submod messagebox
- Added several structs and enum next to it
- Added a messagebox example in examples/ to :
  - allow users to test it quickly
  - have something for the devs to test this feature quickly

Also changed some minor stuff in show_simple_message_box,
but it shouldn't break the current API

AngryLawyer commented on src/sdl2/messagebox.rs:
```diff
@@ -48,3 +101,60 @@ pub fn show_simple_message_box(flags: MessageBoxFlag, title: &str,
     }
 }
+
+/// returns the id of the button triggered
```
This could do with slightly better documentation

AngryLawyer commented on src/sdl2/messagebox.rs:
```diff
@@ -48,3 +101,60 @@ pub fn show_simple_message_box(flags: MessageBoxFlag, title: &str,
 }
+
+/// returns the id of the button triggered
+pub fn show_message_box<'a>(flags:MessageBoxFlag, buttons:&'a [ButtonData], title:&str,
```
Is a button id really the best thing to return from this? My heart tells me what we really want is a reference to one of the borrowed buttons we feed in, but I've never really used message boxes so I'm not all that sure.

Cobrand replied:
I'm fine with both, I just thought that integers would be easier to handle than a potential reference of ButtonData. I think it can lead to lifetime problems too, but I'm not sure. Either way is fine by me.

I think it would be better to stay close to the raw SDL implementation and keep the Result<i32>, but that's just my opinion.

AngryLawyer commented on src/sdl2/messagebox.rs:
```diff
@@ -48,3 +101,60 @@ pub fn show_simple_message_box(flags: MessageBoxFlag, title: &str,
+    use self::ShowMessageError::*;
+    let mut button_id : c_int = 0;
+    let result = unsafe {
+        let title = match CString::new(title) {
```
I don't think you need to allocate new strings here (unless messagebox is mutating these things, in which case wtf?) - Rust strings can be cast into happy happy *const c_chars for SDL2 to use.

Look here for an example:

https://github.com/AngryLawyer/rust-sdl2/blob/648e448cb718a34ce90c3e36a70083b1635fe89a/src/sdl2/hint.rs#L16

AngryLawyer replied:
Oh fair enough

AngryLawyer commented on src/sdl2/messagebox.rs:
```diff
@@ -48,3 +101,60 @@ pub fn show_simple_message_box(flags: MessageBoxFlag, title: &str,
+    -> Result<i32,ShowMessageError> {
+    use self::ShowMessageError::*;
+    let mut button_id : c_int = 0;
+    let result = unsafe {
```
The unsafe block should generally be as small as possible; this is covering large areas of code that don't need to be unsafe.

Cobrand commented:
Bump, returning a `button_id` instead of a reference of a button is required because it can also return -1, which represents the close button (and thus no button at all), meaning it can be returned even if the list of buttons is empty.

As such, I updated the unsafe part and added comments / docs to both show_message_box and show_simple_message_box

AngryLawyer commented:
In which case, surely Some(Button) | None is better?

Cobrand commented:
Bump, I made a new enum instead of Option, making it more explicit that the close button has been pressed, like so : 
```rust
pub enum ClickedButton<'a> {
    CloseButton,
    CustomButton(&'a ButtonData<'a>)
}
```

AngryLawyer commented:
Looks great!
